File: sdk/swift/Tests/SeocheonSDKTests/ClientTests.swift
import XCTest
@testable import SeocheonSDK

final class ClientTests: XCTestCase {

    private func makeConfig() -> SDKConfig {
        return SDKConfig(
            chain: ChainConfig(chainId: "seocheon-1", rpcEndpoint: "http://localhost:26657", grpcEndpoint: "http://localhost:1317"),
            signing: SigningConfig(mode: .direct, mnemonic: "abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon about")
        )
    }

    func testSDKCreation() throws {
        let sdk = try SeocheonSDK(config: makeConfig())
        XCTAssertNotNil(sdk.activity)
        XCTAssertNotNil(sdk.epoch)
        XCTAssertNotNil(sdk.node)
        XCTAssertNotNil(sdk.rewards)
        XCTAssertNotNil(sdk.cosmos)
    }

    func testSDKIsNotConnectedByDefault() throws {
        let sdk = try SeocheonSDK(config: makeConfig())
        XCTAssertFalse(sdk.isConnected())
    }

    func testSDKGetConfig() throws {
        let config = makeConfig()
        let sdk = try SeocheonSDK(config: config)
        let returned = sdk.getConfig()
        XCTAssertEqual(returned.chain.chainId, "seocheon-1")
    }

    func testSDKInvalidConfigThrows() {
        let config = SDKConfig(
            chain: ChainConfig(chainId: "", rpcEndpoint: "", grpcEndpoint: ""),
            signing: SigningConfig(mode: .direct)
        )
        XCTAssertThrowsError(try SeocheonSDK(config: config))
    }

    func testSDKWithMockClient() async throws {
        let config = makeConfig()
        let mockClient = MockChainClient()
        let mockSigner = MockSigner()
        let sdk = SeocheonSDK(config: config, client: mockClient, signer: mockSigner)

        XCTAssertFalse(sdk.isConnected())
        try await sdk.connect()
        XCTAssertTrue(sdk.isConnected())
        await sdk.disconnect()
        XCTAssertFalse(sdk.isConnected())
    }
}

File: sdk/csharp/tests/ClientTests.cs
using Seocheon.Sdk.Errors;
using Seocheon.Sdk.Tests.TestHelpers;
using Xunit;

namespace Seocheon.Sdk.Tests;

public class ClientTests
{
    private static SdkConfig TestConfig() => new()
    {
        Chain = new ChainConfig
        {
            ChainId = "seocheon-test-1",
            RpcEndpoint = "http://localhost:26657",
            GrpcEndpoint = "http://localhost:1317"
        },
        Signing = new SigningConfig
        {
            Mode = SigningMode.Direct,
            Mnemonic = "abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon about"
        }
    };

    [Fact]
    public void CreateSdk_WithMock()
    {
        var client = new MockChainClient();
        var signer = new MockSigner();

        var sdk = new SeocheonSdk(TestConfig(), client, signer);

        Assert.NotNull(sdk.Activity);
        Assert.NotNull(sdk.Epoch);
        Assert.NotNull(sdk.Node);
        Assert.NotNull(sdk.Rewards);
        Assert.NotNull(sdk.Cosmos);
        Assert.False(sdk.IsConnected);
    }

    [Fact]
    public async Task ConnectAndDisconnect()
    {
        var client = new MockChainClient();
        var signer = new MockSigner();

        using var sdk = new SeocheonSdk(TestConfig(), client, signer);

        await sdk.ConnectAsync();
        Assert.True(sdk.IsConnected);

        await sdk.DisconnectAsync();
        Assert.False(sdk.IsConnected);
    }

    [Fact]
    public void InvalidConfig_Throws()
    {
        var config = new SdkConfig
        {
            Chain = new ChainConfig { ChainId = "", RpcEndpoint = "", GrpcEndpoint = "" },
            Signing = new SigningConfig { Mode = SigningMode.Direct }
        };

        Assert.Throws<SdkException>(() => new SeocheonSdk(config));
    }

    [Fact]
    public void Config_Property_ReturnsConfig()
    {
        var config = TestConfig();
        var client = new MockChainClient();
        var signer = new MockSigner();

        var sdk = new SeocheonSdk(config, client, signer);
        Assert.Equal("seocheon-test-1", sdk.Config.Chain.ChainId);
    }

    [Fact]
    public void GetAddress_ReturnsMockAddress()
    {
        var client = new MockChainClient();
        var signer = new MockSigner { Address = "seocheon1myaddress" };

        var sdk = new SeocheonSdk(TestConfig(), client, signer);
        Assert.Equal("seocheon1myaddress", sdk.GetAddress());
    }
}
